backend/test_observer/controllers/reports/reports.py

```diff
@@ -72,10 +107,89 @@ def get_testresults_report(
         .where(Artefact.created_at >= start_date, Artefact.created_at <= end_date)
     )
 
-    filename = "testresults_report.csv"
-    with open(filename, "w") as csvfile:
-        writer = csv.writer(csvfile)
-        writer.writerow(TESTRESULTS_REPORT_COLUMNS)
-        writer.writerows(cursor)
+    return _report_csv_format(
+        "testresults_report.csv",
+        TESTRESULTS_REPORT_COLUMNS,
+        cursor,
+    )
 
-    return filename
+
+def _get_test_execution_reports_query(
+    columns: list[InstrumentedAttribute],
+    include_test_events: bool,
+    start_date: datetime,
+    end_date: datetime,
+) -> Select:
+    """
+    Builds the query that retrieves the test executions based on the parameters set
+    """
+    latest_builds = queries.latest_artefact_builds.subquery()
+    base_query = (
+        select(*columns)
+        .join_from(TestExecution, Environment)
+        .join_from(TestExecution, latest_builds)
+        .join_from(latest_builds, Artefact)
+        .join_from(Artefact, Stage)
+        .join_from(Stage, Family)
+        .join(
+            ArtefactBuildEnvironmentReview,
+            (ArtefactBuildEnvironmentReview.artefact_build_id == latest_builds.c.id)
+            & (ArtefactBuildEnvironmentReview.environment_id == Environment.id),
+        )
+        .where(
+            TestExecution.created_at >= start_date, TestExecution.created_at <= end_date
+        )
+    )
+
+    if include_test_events:
+        return base_query.outerjoin(
+            TestEvent, TestExecution.id == TestEvent.test_execution_id
+        ).group_by(*TEST_EXECUTIONS_REPORT_COLUMNS)
+
+    return base_query
+
+
+@router.get("/test-executions", response_class=FileResponse)
+def get_test_execution_reports(
+    start_date: datetime = datetime.min,
+    end_date: datetime | None = None,
+    include_test_events: bool = False,
+    db: Session = Depends(get_db),
+):
+    """
+    Returns a csv report detailing all test executions within a given date range.
+    Together with their artefact and environment details in csv format.
+    """
+
+    if end_date is None:
+        end_date = datetime.now()
+
+    columns = [*TEST_EXECUTIONS_REPORT_COLUMNS]
+    column_names = [str(column) for column in TEST_EXECUTIONS_REPORT_COLUMNS]
+    if include_test_events:
+        columns.extend(
+            [
+                func.array_agg(
+                    aggregate_order_by(TestEvent.event_name, TestEvent.timestamp)
+                ),
+                func.array_agg(
+                    aggregate_order_by(
+                        func.to_char(TestEvent.timestamp, "YYYY-MM-DD HH24:MI:SS:MS"),
+                        TestEvent.timestamp,
+                    ),
+                ),
+                func.array_agg(
+                    aggregate_order_by(TestEvent.detail, TestEvent.timestamp)
+                ),
+            ]
```

Comment:
shouldn't the aggregate order be on timestamp always? we care about the order in which the events came in

Reply:
Yes, they should be, also, now I replaced with a dictionary instead of tuple, so we don't need the ordering.